canoe/sp1-cc/client/src/main.rs

```diff
@@ -14,46 +14,68 @@ pub fn main() {
-    // V2 will be deprecated once router is released, will remove V2 call then
-    let call = match CertVerifierCall::build(&canoe_input.altda_commitment) {
-        CertVerifierCall::V2(call) => {
-            ContractInput::new_call(verifier_address, Address::default(), call)
-        }
-        CertVerifierCall::Router(call) => {
-            ContractInput::new_call(verifier_address, Address::default(), call)
-        }
-    };
-
-    let public_vals = executor.execute(call).unwrap();
-
-    // empricially if the function reverts, the output is empty, the guest code abort when evm revert takes place
-    let returns = Bool::abi_decode(&public_vals.contractOutput).expect("deserialize returns");
-
-    let rlp_bytes = canoe_input.altda_commitment.to_rlp_bytes();
-
     let chain_sepc: ChainSpec = executor
         .genesis
         .try_into()
         .expect("convert sp1 genesis into chain spec");
 
-    let journal = Journal {
-        certVerifierAddress: verifier_address,
-        input: rlp_bytes.into(),
-        blockhash: public_vals.anchorHash,
-        output: returns,
-        l1ChainId: chain_sepc.chain.id(),
-    };
+    // journals is a concatenation of serialized proof journal for each canoe input. The whole serialized vec is
+    // reproduced during the verification procedures
+    let mut journals: Vec<u8> = vec![];
+    // executes all calls, then combines and commits all journals
+    for canoe_input in canoe_inputs.iter() {
+        // TODO, are there no better way to reduce this duplicate code.
+        // known constraint, new_call takes SolCall trait, which is Sized so not dyn trait
+        // also impl SolCall seems too much
+        // V2 will be deprecated once router is released, will remove V2 call then
+        let call = match CertVerifierCall::build(&canoe_input.altda_commitment) {
+            CertVerifierCall::V2(call) => {
+                ContractInput::new_call(canoe_input.verifier_address, Address::default(), call)
+            }
+            CertVerifierCall::Router(call) => {
+                ContractInput::new_call(canoe_input.verifier_address, Address::default(), call)
+            }
+        };
+
+        let public_vals = executor.execute(call).unwrap();
+
+        // empricially if the function reverts, the output is empty, the guest code abort when evm revert takes place
+        let returns = Bool::abi_decode(&public_vals.contractOutput).expect("deserialize returns");
+
+        // TODO might be using a better serialization format
+        let rlp_bytes = canoe_input.altda_commitment.to_rlp_bytes();
+
+        assert!(public_vals.anchorHash == l1_head_block_hash);
+
+        let journal = Journal {
+            certVerifierAddress: canoe_input.verifier_address,
+            input: rlp_bytes.into(),
+            blockhash: public_vals.anchorHash,
+            output: returns,
+            l1ChainId: chain_sepc.chain.id(),
+        };
+        journals.extend(journal.abi_encode());
```

Comment:
how do you deserialize the concatenation? You don't know how long each abi encoded journal is right..?

Reply:
Gotcha I see. Maybe add a comment at least explaining what you outlined above (that this never gets deserialized)?